// @pierre/precision-diffs/dist/utils/createTransformerWithState.js
import { convertLine, createAnnotationElement, createEmptyRowBuffer, findCodeElement } from "./hast_utils.js";
import { transformerStyleToClass } from "@shikijs/transformers";

//#region src/utils/createTransformerWithState.ts
function createTransformerWithState({ disableLineNumbers, useCSSClasses }) {
	const state = {
		lineInfo: {},
		decorations: [],
		disableLineNumbers
	};
	const transformers = [{
		line(node) {
			delete node.properties.class;
			return node;
		},
		pre(pre) {
			const code = findCodeElement(pre);
			const children = [];
			if (code != null) {
				let index = 1;
				for (const node of code.children) {
					if (node.type !== "element") continue;
					if (index === 1 && state.lineInfo[0]?.spans != null) for (const span of state.lineInfo[0]?.spans ?? []) if (span.type === "gap") children.push(createEmptyRowBuffer(span.rows));
					else children.push(createAnnotationElement(span));
					children.push(convertLine(node, index, state));
					const lineInfo = state.lineInfo[index];
					if (lineInfo?.spans != null) for (const span of lineInfo.spans) if (span.type === "gap") children.push(createEmptyRowBuffer(span.rows));
					else children.push(createAnnotationElement(span));
					index++;
				}
				code.children = children;
			}
			return pre;
		}
	}];
	if (useCSSClasses) transformers.push(tokenStyleNormalizer, toClass);
	return {
		state,
		transformers,
		toClass
	};
}
const toClass = transformerStyleToClass({ classPrefix: "hl-" });
const tokenStyleNormalizer = {
	name: "token-style-normalizer",
	tokens(lines) {
		for (const line of lines) for (const token of line) {
			if (token.htmlStyle != null) continue;
			const style = {};
			if (token.color != null) style.color = token.color;
			if (token.bgColor != null) style["background-color"] = token.bgColor;
			if (token.fontStyle != null && token.fontStyle !== 0) {
				if ((token.fontStyle & 1) !== 0) style["font-style"] = "italic";
				if ((token.fontStyle & 2) !== 0) style["font-weight"] = "bold";
				if ((token.fontStyle & 4) !== 0) style["text-decoration"] = "underline";
			}
			if (Object.keys(style).length > 0) token.htmlStyle = style;
		}
	}
};

//#endregion
export { createTransformerWithState };
