// @pierre/precision-diffs/dist/utils/hast_utils.js
import { DEFAULT_THEMES, HEADER_METADATA_SLOT_ID } from "../constants.js";
import { getHighlighterThemeStyles } from "./getHighlighterThemeStyles.js";
import { getIconForType } from "./getIconForType.js";

//#region src/utils/hast_utils.ts
function createTextNode(value) {
	return {
		type: "text",
		value
	};
}
function createHastElement({ tagName, children = [], properties = {} }) {
	return {
		type: "element",
		tagName,
		properties,
		children
	};
}
function createSeparator({ type, content, expandIndex, slotName, isFirstHunk, isLastHunk }) {
	const children = [];
	if (type === "metadata" && content != null) children.push(createHastElement({
		tagName: "div",
		children: [createTextNode(content)],
		properties: { "data-separator-content": "" }
	}));
	if (type === "line-info" && content != null) {
		const contentChildren = [createTextNode(content)];
		if (expandIndex != null) contentChildren.unshift(createIcon({ name: "pjs-icon-chevrons-narrow" }));
		children.push(createHastElement({
			tagName: "div",
			children: contentChildren,
			properties: { "data-separator-content": "" }
		}));
	}
	if (type === "custom" && slotName != null) children.push(createHastElement({
		tagName: "slot",
		properties: { name: slotName }
	}));
	return createHastElement({
		tagName: "div",
		children,
		properties: {
			"data-separator": children.length === 0 ? "" : type,
			"data-expand-index": expandIndex,
			"data-separator-first": isFirstHunk ? "" : void 0,
			"data-separator-last": isLastHunk ? "" : void 0
		}
	});
}
function createIcon({ name, width = 16, height = 16, properties }) {
	return createHastElement({
		tagName: "svg",
		properties: {
			width,
			height,
			viewBox: "0 0 16 16",
			...properties
		},
		children: [createHastElement({
			tagName: "use",
			properties: { href: `#${name.replace(/^#/, "")}` }
		})]
	});
}
function createAnnotationElement(span) {
	return createHastElement({
		tagName: "div",
		children: [createHastElement({
			tagName: "div",
			children: span.annotations?.map((slotId) => createHastElement({
				tagName: "slot",
				properties: { name: slotId }
			})),
			properties: { "data-annotation-content": "" }
		})],
		properties: { "data-line-annotation": `${span.hunkIndex},${span.lineIndex}` }
	});
}
function createEmptyRowBuffer(size) {
	return createHastElement({
		tagName: "div",
		properties: {
			"data-buffer": "",
			style: `grid-row: span ${size};min-height:calc(${size} * 1lh)`
		}
	});
}
function findCodeElement(nodes) {
	let firstChild = nodes.children[0];
	while (firstChild != null) {
		if (firstChild.type === "element" && firstChild.tagName === "code") return firstChild;
		if ("children" in firstChild) firstChild = firstChild.children[0];
		else firstChild = null;
	}
}
function convertLine(node, line, state) {
	const lineInfo = state.lineInfo[line];
	if (lineInfo == null) {
		console.error({
			node,
			line,
			state
		});
		throw new Error(`convertLine: line ${line}, contains no state.lineInfo`);
	}
	node.tagName = "span";
	node.properties["data-column-content"] = "";
	if (lineInfo.metadataContent != null) node.children.push(createHastElement({
		tagName: "span",
		children: [createTextNode(lineInfo.metadataContent)],
		properties: { "data-no-newline": "" }
	}));
	else if (node.children.length === 0) node.children.push(createTextNode("\n"));
	const children = [node];
	if (!state.disableLineNumbers) children.unshift(createHastElement({
		tagName: "span",
		children: lineInfo.metadataContent == null ? [{
			type: "text",
			value: `${lineInfo.lineNumber}`
		}] : [],
		properties: { "data-column-number": "" }
	}));
	return createHastElement({
		tagName: "div",
		children,
		properties: {
			"data-line": lineInfo.metadataContent == null ? `${lineInfo.lineNumber}` : "",
			"data-line-type": lineInfo.type
		}
	});
}
function createPreWrapperProperties({ diffIndicators = "bars", disableBackground = false, highlighter, overflow = "scroll", split, theme = DEFAULT_THEMES, themeType = "system", totalLines }) {
	const properties = {
		"data-pjs": "",
		"data-type": split ? "split" : "file",
		"data-overflow": overflow,
		style: getHighlighterThemeStyles({
			theme,
			highlighter
		}),
		tabIndex: 0
	};
	properties.style += `--pjs-min-number-column-width-default:${`${totalLines}`.length}ch;`;
	if (typeof theme === "string" && themeType !== "system") properties["data-theme-type"] = themeType;
	else if (typeof theme === "string") properties["data-theme-type"] = highlighter.getTheme(theme).type;
	switch (diffIndicators) {
		case "bars":
		case "classic":
			properties["data-indicators"] = diffIndicators;
			break;
	}
	if (!disableBackground) properties["data-background"] = "";
	return properties;
}
function createFileHeaderElement({ fileOrDiff, theme, highlighter, prefix, themeType = "system" }) {
	const fileDiff = "type" in fileOrDiff ? fileOrDiff : void 0;
	const properties = {
		"data-pjs-header": "",
		style: getHighlighterThemeStyles({
			theme,
			highlighter,
			prefix
		})
	};
	if (fileDiff != null) properties["data-change-type"] = fileDiff.type;
	if (typeof theme === "string") properties["data-theme-type"] = highlighter.getTheme(theme).type;
	else if (themeType !== "system") properties["data-theme-type"] = themeType;
	return createHastElement({
		tagName: "div",
		children: [createHeaderElement({
			name: fileOrDiff.name,
			prevName: "prevName" in fileOrDiff ? fileOrDiff.prevName : void 0,
			iconType: fileDiff?.type ?? "file"
		}), createMetadataElement(fileDiff)],
		properties
	});
}
function createHeaderElement({ name, prevName, iconType }) {
	const children = [createIcon({
		name: getIconForType(iconType),
		properties: { "data-change-icon": iconType }
	})];
	if (prevName != null) {
		children.push(createHastElement({
			tagName: "div",
			children: [createTextNode(prevName)],
			properties: { "data-prev-name": "" }
		}));
		children.push(createIcon({
			name: "pjs-icon-arrow",
			properties: { "data-rename-icon": "" }
		}));
	}
	children.push(createHastElement({
		tagName: "div",
		children: [createTextNode(name)],
		properties: { "data-title": "" }
	}));
	return createHastElement({
		tagName: "div",
		children,
		properties: { "data-header-content": "" }
	});
}
function createMetadataElement(fileDiff) {
	const children = [];
	if (fileDiff != null) {
		let additions = 0;
		let deletions = 0;
		for (const hunk of fileDiff.hunks) for (const line of hunk.hunkContent ?? []) if (line.startsWith("+")) additions++;
		else if (line.startsWith("-")) deletions++;
		if (deletions > 0) children.push(createHastElement({
			tagName: "span",
			children: [createTextNode(`-${deletions}`)],
			properties: { "data-deletions-count": "" }
		}));
		if (additions > 0) children.push(createHastElement({
			tagName: "span",
			children: [createTextNode(`+${additions}`)],
			properties: { "data-additions-count": "" }
		}));
		if (deletions === 0 && additions === 0) children.push(createHastElement({
			tagName: "span",
			children: [createTextNode("No diff")]
		}));
	}
	children.push(createHastElement({
		tagName: "slot",
		properties: { name: HEADER_METADATA_SLOT_ID }
	}));
	return createHastElement({
		tagName: "div",
		children,
		properties: { "data-metadata": "" }
	});
}

//#endregion
export { convertLine, createAnnotationElement, createEmptyRowBuffer, createFileHeaderElement, createHastElement, createIcon, createPreWrapperProperties, createSeparator, createTextNode, findCodeElement };
