// @pierre/precision-diffs/dist/utils/parseDiffDecorations.js
import { diffChars, diffWordsWithSpace } from "diff";

//#region src/utils/parseDiffDecorations.ts
function parseDecorations({ diffGroups, disableDecorations = false, lineDiffType, maxLineDiffLength, diffStyle }) {
	const unified = diffStyle === "unified";
	const unifiedDecorations = [];
	const additionDecorations = [];
	const deletionDecorations = [];
	if (disableDecorations || lineDiffType === "none") return {
		unifiedDecorations,
		deletionDecorations,
		additionDecorations
	};
	for (const group of diffGroups) {
		const len = Math.min(group.additionLines.length, group.deletionLines.length);
		for (let i = 0; i < len; i++) {
			const deletionLine = group.deletionLines[i];
			const additionLine = group.additionLines[i];
			if (deletionLine == null || additionLine == null) break;
			if (deletionLine.length > maxLineDiffLength || additionLine.length > maxLineDiffLength) continue;
			const lineDiff = lineDiffType === "char" ? diffChars(deletionLine, additionLine) : diffWordsWithSpace(deletionLine, additionLine);
			const deletionSpans = [];
			const additionSpans = [];
			const enableJoin = lineDiffType === "word-alt";
			for (const item of lineDiff) if (!item.added && !item.removed) {
				pushOrJoinSpan({
					item,
					arr: deletionSpans,
					enableJoin,
					isNeutral: true
				});
				pushOrJoinSpan({
					item,
					arr: additionSpans,
					enableJoin,
					isNeutral: true
				});
			} else if (item.removed) pushOrJoinSpan({
				item,
				arr: deletionSpans,
				enableJoin
			});
			else pushOrJoinSpan({
				item,
				arr: additionSpans,
				enableJoin
			});
			let spanIndex = 0;
			for (const span of additionSpans) {
				if (span[0] === 1) (unified ? unifiedDecorations : additionDecorations).push(createDiffSpanDecoration({
					line: group.additionStartIndex + i,
					spanStart: spanIndex,
					spanLength: span[1].length
				}));
				spanIndex += span[1].length;
			}
			spanIndex = 0;
			for (const span of deletionSpans) {
				if (span[0] === 1) (unified ? unifiedDecorations : deletionDecorations).push(createDiffSpanDecoration({
					line: group.deletionStartIndex + i,
					spanStart: spanIndex,
					spanLength: span[1].length
				}));
				spanIndex += span[1].length;
			}
		}
	}
	return {
		unifiedDecorations,
		deletionDecorations,
		additionDecorations
	};
}
function createDiffSpanDecoration({ line, spanStart, spanLength }) {
	return {
		start: {
			line,
			character: spanStart
		},
		end: {
			line,
			character: spanStart + spanLength
		},
		properties: { "data-diff-span": "" },
		alwaysWrap: true
	};
}
function pushOrJoinSpan({ item, arr, enableJoin, isNeutral = false }) {
	const lastItem = arr[arr.length - 1];
	if (lastItem == null || item.value === "\n" || !enableJoin) {
		arr.push([isNeutral ? 0 : 1, item.value]);
		return;
	}
	const isLastItemNeutral = lastItem[0] === 0;
	if (isNeutral === isLastItemNeutral || isNeutral && item.value.length === 1 && !isLastItemNeutral) {
		lastItem[1] += item.value;
		return;
	}
	arr.push([isNeutral ? 0 : 1, item.value]);
}

//#endregion
export { parseDecorations };
