// @pierre/precision-diffs/dist/FileStream.js
import { getSharedHighlighter } from "./SharedHighlighter.js";
import { DEFAULT_THEMES } from "./constants.js";
import { formatCSSVariablePrefix } from "./utils/formatCSSVariablePrefix.js";
import { getHighlighterOptions } from "./utils/getHighlighterOptions.js";
import { createCodeNode, createRow, createSpanFromToken, setWrapperProps } from "./utils/html_render_utils.js";
import { queueRender } from "./UniversalRenderer.js";
import { CodeToTokenTransformStream } from "./shiki-stream/stream.js";

//#region src/FileStream.ts
var FileStream = class {
	highlighter;
	stream;
	fileContainer;
	pre;
	code;
	constructor(options = { theme: DEFAULT_THEMES }) {
		this.options = options;
		this.currentLineIndex = this.options.startingLineIndex ?? 1;
	}
	cleanUp() {
		this.stream?.cancel();
	}
	setThemeType(themeType) {
		if ((this.options.themeType ?? "system") === themeType) return;
		this.options = {
			...this.options,
			themeType
		};
		if (this.pre != null) switch (themeType) {
			case "system":
				delete this.pre.dataset.themeType;
				break;
			case "light":
			case "dark":
				this.pre.dataset.themeType = themeType;
				break;
		}
	}
	async initializeHighlighter() {
		this.highlighter = await getSharedHighlighter(getHighlighterOptions(this.options.lang, this.options));
		return this.highlighter;
	}
	queuedSetupArgs;
	async setup(_source, _wrapper) {
		const isSettingUp = this.queuedSetupArgs != null;
		this.queuedSetupArgs = [_source, _wrapper];
		if (isSettingUp) return;
		this.highlighter ??= await this.initializeHighlighter();
		const [source, wrapper] = this.queuedSetupArgs;
		this.queuedSetupArgs = void 0;
		const stream = source;
		this.setupStream(stream, wrapper, this.highlighter);
	}
	setupStream(stream, wrapper, highlighter) {
		const { theme = DEFAULT_THEMES, overflow = "scroll", themeType = "system" } = this.options;
		const fileContainer = this.getOrCreateFileContainer();
		if (fileContainer.parentElement == null) wrapper.appendChild(fileContainer);
		this.pre ??= document.createElement("pre");
		if (this.pre.parentElement == null) fileContainer.shadowRoot?.appendChild(this.pre);
		const pre = setWrapperProps({
			pre: this.pre,
			split: false,
			theme,
			highlighter,
			wrap: overflow === "wrap",
			themeType,
			diffIndicators: "none",
			disableBackground: true,
			totalLines: 0
		});
		pre.innerHTML = "";
		this.pre = pre;
		this.code = createCodeNode({ pre });
		if (this.stream != null) this.stream.cancel();
		const { onStreamStart, onStreamClose, onStreamAbort } = this.options;
		this.stream = stream;
		this.stream.pipeThrough(typeof theme === "string" ? new CodeToTokenTransformStream({
			...this.options,
			theme,
			highlighter,
			allowRecalls: true,
			defaultColor: false,
			cssVariablePrefix: formatCSSVariablePrefix()
		}) : new CodeToTokenTransformStream({
			...this.options,
			themes: theme,
			highlighter,
			allowRecalls: true,
			defaultColor: false,
			cssVariablePrefix: formatCSSVariablePrefix()
		})).pipeTo(new WritableStream({
			start(controller) {
				onStreamStart?.(controller);
			},
			close() {
				onStreamClose?.();
			},
			abort(reason) {
				onStreamAbort?.(reason);
			},
			write: this.handleWrite
		}));
	}
	queuedTokens = [];
	handleWrite = (token) => {
		if ("recall" in token && this.queuedTokens.length >= token.recall) this.queuedTokens.length = this.queuedTokens.length - token.recall;
		else this.queuedTokens.push(token);
		queueRender(this.render);
		this.options.onStreamWrite?.(token);
	};
	currentLineIndex;
	currentLineElement;
	render = () => {
		this.options.onPreRender?.(this);
		const linesToAppend = [];
		for (const token of this.queuedTokens) if ("recall" in token) {
			if (this.currentLineElement == null) throw new Error("FileStream.render: no current line element, shouldnt be possible to get here");
			if (token.recall > this.currentLineElement.childNodes.length) throw new Error(`FileStream.render: Token recall exceed the current line, there's probably a bug...`);
			for (let i = 0; i < token.recall; i++) this.currentLineElement.lastChild?.remove();
		} else {
			const span = createSpanFromToken(token);
			if (this.currentLineElement == null) linesToAppend.push(this.createLine());
			this.currentLineElement?.appendChild(span);
			if (token.content === "\n") {
				this.currentLineIndex++;
				linesToAppend.push(this.createLine());
			}
		}
		for (const line of linesToAppend) this.code?.appendChild(line);
		this.queuedTokens.length = 0;
		this.options.onPostRender?.(this);
	};
	createLine() {
		const { row, content } = createRow(this.currentLineIndex);
		this.currentLineElement = content;
		return row;
	}
	getOrCreateFileContainer(fileContainer) {
		if (fileContainer != null && fileContainer === this.fileContainer || fileContainer == null && this.fileContainer != null) return this.fileContainer;
		this.fileContainer = fileContainer ?? document.createElement("file-diff");
		return this.fileContainer;
	}
};

//#endregion
export { FileStream };
