// @pierre/precision-diffs/dist/FileHeaderRenderer.js
import { getSharedHighlighter, hasLoadedThemes } from "./SharedHighlighter.js";
import { DEFAULT_THEMES } from "./constants.js";
import { createFileHeaderElement } from "./utils/hast_utils.js";
import { getThemes } from "./utils/getThemes.js";
import { getHighlighterOptions } from "./utils/getHighlighterOptions.js";
import { toHtml } from "hast-util-to-html";

//#region src/FileHeaderRenderer.ts
var FileHeaderRenderer = class {
	highlighter;
	constructor(options = { theme: DEFAULT_THEMES }) {
		this.options = options;
	}
	cleanUp() {
		this.highlighter = void 0;
		this.queuedRenderFileOrDiff = void 0;
		this.queuedRender = void 0;
	}
	mergeOptions(options) {
		this.options = {
			...this.options,
			...options
		};
	}
	setOptions(options) {
		this.options = options;
	}
	setThemeType(themeType) {
		if ((this.options.themeType ?? "system") === themeType) return;
		this.mergeOptions({ themeType });
	}
	async initializeHighlighter() {
		this.highlighter = await getSharedHighlighter(getHighlighterOptions(void 0, this.options));
		return this.highlighter;
	}
	fileOrDiff;
	queuedRenderFileOrDiff;
	queuedRender;
	async render(fileOrDiff) {
		this.queuedRenderFileOrDiff = fileOrDiff;
		if (this.queuedRender != null) return this.queuedRender;
		this.queuedRender = (async () => {
			if (!hasLoadedThemes(getThemes(this.options.theme))) this.highlighter = void 0;
			this.highlighter ??= await this.initializeHighlighter();
			if (this.queuedRenderFileOrDiff == null) return;
			return this.renderHeader(this.queuedRenderFileOrDiff, this.highlighter);
		})();
		const result = await this.queuedRender;
		this.queuedRenderFileOrDiff = void 0;
		this.queuedRender = void 0;
		return result;
	}
	renderHeader(fileOrDiff, highlighter) {
		this.fileOrDiff = fileOrDiff;
		return createFileHeaderElement({
			...this.options,
			fileOrDiff,
			highlighter
		});
	}
	renderResultToHTML(element) {
		return toHtml(element);
	}
};

//#endregion
export { FileHeaderRenderer };
