// @pierre/precision-diffs/dist/ResizeManager.js
//#region src/ResizeManager.ts
var ResizeManager = class {
	observedNodes = /* @__PURE__ */ new Map();
	cleanUp() {
		this.resizeObserver?.disconnect();
		this.observedNodes.clear();
	}
	resizeObserver;
	setup(pre) {
		this.cleanUp();
		const annotationElements = pre.querySelectorAll("[data-line-annotation*=\",\"]");
		this.resizeObserver ??= new ResizeObserver(this.handleResizeObserver);
		const codeElements = pre.querySelectorAll("code");
		for (const codeElement of codeElements) {
			let numberElement = codeElement.querySelector("[data-column-number]");
			if (!(numberElement instanceof HTMLElement)) numberElement = null;
			const item = {
				type: "code",
				codeElement,
				numberElement,
				codeWidth: "auto",
				numberWidth: 0
			};
			this.observedNodes.set(codeElement, item);
			this.resizeObserver.observe(codeElement);
			if (numberElement != null) {
				this.observedNodes.set(numberElement, item);
				this.resizeObserver.observe(numberElement);
			}
		}
		if (codeElements.length <= 1) return;
		const elementMap = /* @__PURE__ */ new Map();
		for (const element of annotationElements) {
			if (!(element instanceof HTMLElement)) continue;
			const { lineAnnotation = "" } = element.dataset;
			if (!/^\d+,\d+$/.test(lineAnnotation)) {
				console.error("DiffFileRenderer.setupResizeObserver: Invalid element or annotation", {
					lineAnnotation,
					element
				});
				continue;
			}
			let pairs = elementMap.get(lineAnnotation);
			if (pairs == null) {
				pairs = [];
				elementMap.set(lineAnnotation, pairs);
			}
			pairs.push(element);
		}
		for (const [key, pair] of elementMap) {
			if (pair.length !== 2) {
				console.error("DiffFileRenderer.setupResizeObserver: Bad Pair", key, pair);
				continue;
			}
			const [container1, container2] = pair;
			const child1 = container1.firstElementChild;
			const child2 = container2.firstElementChild;
			if (!(container1 instanceof HTMLElement) || !(container2 instanceof HTMLElement) || !(child1 instanceof HTMLElement) || !(child2 instanceof HTMLElement)) continue;
			const item = {
				type: "annotations",
				column1: {
					container: container1,
					child: child1,
					childHeight: 0
				},
				column2: {
					container: container2,
					child: child2,
					childHeight: 0
				},
				currentHeight: "auto"
			};
			this.observedNodes.set(child1, item);
			this.observedNodes.set(child2, item);
			this.resizeObserver.observe(child1);
			this.resizeObserver.observe(child2);
		}
	}
	handleResizeObserver = (entries) => {
		for (const entry of entries) {
			const { target, borderBoxSize } = entry;
			if (!(target instanceof HTMLElement)) {
				console.error("FileDiff.handleResizeObserver: Invalid element for ResizeObserver", entry);
				continue;
			}
			const item = this.observedNodes.get(target);
			if (item == null) {
				console.error("FileDiff.handleResizeObserver: Not a valid observed node", entry);
				continue;
			}
			const specs = borderBoxSize[0];
			if (item.type === "annotations") {
				const column = (() => {
					if (target === item.column1.child) return item.column1;
					if (target === item.column2.child) return item.column2;
				})();
				if (column == null) {
					console.error(`FileDiff.handleResizeObserver: Couldn't find a column for`, {
						item,
						target
					});
					continue;
				}
				column.childHeight = specs.blockSize;
				const newHeight = Math.max(item.column1.childHeight, item.column2.childHeight);
				if (newHeight !== item.currentHeight) {
					item.currentHeight = Math.max(newHeight, 0);
					item.column1.container.style.setProperty("--pjs-annotation-min-height", `${item.currentHeight}px`);
					item.column2.container.style.setProperty("--pjs-annotation-min-height", `${item.currentHeight}px`);
				}
			} else if (item.type === "code") {
				if (target === item.codeElement) {
					if (specs.inlineSize !== item.codeWidth) {
						item.codeWidth = specs.inlineSize;
						item.codeElement.style.setProperty("--pjs-column-content-width", `${Math.max(item.codeWidth - item.numberWidth, 0)}px`);
						item.codeElement.style.setProperty("--pjs-column-width", `${item.codeWidth}px`);
					}
				} else if (target === item.numberElement) {
					if (specs.inlineSize !== item.numberWidth) {
						item.numberWidth = specs.inlineSize;
						item.codeElement.style.setProperty("--pjs-column-number-width", `${item.numberWidth}px`);
						if (item.codeWidth !== "auto") item.codeElement.style.setProperty("--pjs-column-content-width", `${Math.max(item.codeWidth - item.numberWidth, 0)}px`);
					}
				}
			}
		}
	};
};

//#endregion
export { ResizeManager };
