// @pierre/precision-diffs/dist/SharedHighlighter.js
import { createHighlighter, createJavaScriptRegexEngine, createOnigurumaEngine, loadWasm } from "shiki";

//#region src/SharedHighlighter.ts
let highlighter;
const loadedThemes = /* @__PURE__ */ new Map();
const loadedLanguages = /* @__PURE__ */ new Map();
async function getSharedHighlighter({ themes, langs, preferWasmHighlighter = false }) {
	if (isHighlighterNull(highlighter)) {
		highlighter = new Promise((resolve) => {
			(preferWasmHighlighter ? loadWasm(import("shiki/wasm")) : Promise.resolve()).then(() => {
				const themesToLoad = [];
				for (const theme of themes) {
					loadedThemes.set(theme, true);
					const customTheme = CustomThemes.get(theme);
					if (customTheme != null) themesToLoad.push(customTheme());
					else themesToLoad.push(theme);
				}
				for (const language of langs) loadedLanguages.set(language, true);
				loadedLanguages.set("text", true);
				return createHighlighter({
					themes: themesToLoad,
					langs: [...langs, "text"],
					engine: preferWasmHighlighter ? createOnigurumaEngine() : createJavaScriptRegexEngine()
				});
			}).then((instance$1) => {
				highlighter = instance$1;
				resolve(instance$1);
			});
		});
		return highlighter;
	}
	const instance = isHighlighterLoading(highlighter) ? await highlighter : highlighter;
	const loaders = [];
	for (const language of langs) {
		const loadedOrLoading = loadedLanguages.get(language);
		if (loadedOrLoading == null) {
			const promise = instance.loadLanguage(language).then(() => void loadedLanguages.set(language, true));
			loadedLanguages.set(language, promise);
			loaders.push(promise);
		} else if (loadedOrLoading !== true) loaders.push(loadedOrLoading);
	}
	for (const themeName of themes) {
		const loadedOrLoading = loadedThemes.get(themeName);
		if (loadedOrLoading == null) {
			const customTheme = CustomThemes.get(themeName);
			const promise = instance.loadTheme(customTheme != null ? customTheme() : themeName).then(() => void loadedThemes.set(themeName, true));
			loadedThemes.set(themeName, promise);
			loaders.push(promise);
		} else if (loadedOrLoading !== true) loaders.push(loadedOrLoading);
	}
	if (loaders.length > 0) await Promise.all(loaders);
	return instance;
}
function hasLoadedThemes(themes) {
	for (const theme of themes) {
		if (loadedThemes.get(theme) === true) continue;
		return false;
	}
	return true;
}
function hasLoadedLanguage(lang) {
	return loadedLanguages.get(lang) === true;
}
function isHighlighterLoaded(h = highlighter) {
	return h != null && !("then" in h);
}
function isHighlighterLoading(h = highlighter) {
	return h != null && "then" in h;
}
function isHighlighterNull(h = highlighter) {
	return h == null;
}
async function preloadHighlighter(options) {
	await getSharedHighlighter(options);
}
async function disposeHighlighter() {
	if (highlighter == null) return;
	(await highlighter).dispose();
	loadedThemes.clear();
	loadedLanguages.clear();
	highlighter = void 0;
}
const CustomThemes = /* @__PURE__ */ new Map();
function registerCustomTheme(themeName, loader) {
	if (CustomThemes.has(themeName)) {
		console.error("SharedHighlight.registerCustomTheme: theme name already registered", themeName);
		return;
	}
	CustomThemes.set(themeName, loader);
}
registerCustomTheme("pierre-dark", () => import("./themes/pierre-dark.js"));
registerCustomTheme("pierre-light", () => import("./themes/pierre-light.js"));

//#endregion
export { disposeHighlighter, getSharedHighlighter, hasLoadedLanguage, hasLoadedThemes, isHighlighterLoaded, isHighlighterLoading, isHighlighterNull, preloadHighlighter, registerCustomTheme };
