// @pierre/precision-diffs/dist/ssr/FileDiffServer.js
import { createHastElement, createTextNode } from "../utils/hast_utils.js";
import { DiffHunksRenderer } from "../DiffHunksRenderer.js";
import { FileHeaderRenderer } from "../FileHeaderRenderer.js";
import { FileRenderer } from "../FileRenderer.js";
import style_default from "../style.js";
import { SVGSpriteSheet } from "../sprite.js";
import { parseDiffFromFile } from "../utils/parseDiffFromFile.js";
import { toHtml } from "hast-util-to-html";

//#region src/ssr/FileDiffServer.ts
async function preloadFile({ file, options, annotations }) {
	const fileRenderer = new FileRenderer(options);
	const fileHeader = new FileHeaderRenderer(options);
	if (annotations !== void 0 && annotations.length > 0) fileRenderer.setLineAnnotations(annotations);
	const [fileResult, headerResult] = await Promise.all([fileRenderer.render(file, true), fileHeader.render(file)]);
	if (fileResult == null) throw new Error("Failed to render file diff");
	const children = [createHastElement({
		tagName: "style",
		children: [createTextNode(`@layer base, theme, unsafe;
    @layer base {
      ${style_default}
    }
    @layer theme {
      ${fileResult.css}
    }
    @layer unsafe {
      ${options?.unsafeCSS ?? ""}
    }`)]
	})];
	if (headerResult != null) children.push(headerResult);
	const code = fileRenderer.renderFullAST(fileResult);
	code.properties["data-dehydrated"] = "";
	children.push(code);
	return {
		file,
		options,
		annotations,
		prerenderedHTML: `${SVGSpriteSheet}${toHtml(children)}`
	};
}
async function preloadFileDiff({ oldFile, newFile, options, annotations }) {
	const fileDiff = parseDiffFromFile(oldFile, newFile);
	const diffHunksRenderer = new DiffHunksRenderer({
		...options,
		hunkSeparators: typeof options?.hunkSeparators === "function" ? "custom" : options?.hunkSeparators,
		useCSSClasses: true
	});
	const fileHeader = new FileHeaderRenderer(options);
	if (annotations !== void 0 && annotations.length > 0) diffHunksRenderer.setLineAnnotations(annotations);
	const [hunkResult, headerResult] = await Promise.all([diffHunksRenderer.render(fileDiff), fileHeader.render(fileDiff)]);
	if (hunkResult == null) throw new Error("Failed to render file diff");
	const children = [createHastElement({
		tagName: "style",
		children: [createTextNode(`@layer base, theme, unsafe;
    @layer base {
      ${style_default}
    }
    @layer theme {
      ${hunkResult.css}
    }
    @layer unsafe {
      ${options?.unsafeCSS ?? ""}
    }`)]
	})];
	if (headerResult != null) children.push(headerResult);
	const code = diffHunksRenderer.renderFullAST(hunkResult);
	code.properties["data-dehydrated"] = "";
	children.push(code);
	return {
		oldFile,
		newFile,
		options,
		annotations,
		prerenderedHTML: `${SVGSpriteSheet}${toHtml(children)}`
	};
}

//#endregion
export { preloadFile, preloadFileDiff };
